src/ILCompiler.Compiler/src/Compiler/MultiFileCompilationModuleGroup.cs

```diff
@@ -81,29 +81,4 @@ public override bool ShouldProduceFullType(TypeDesc type)
             return ConstructedEETypeNode.CreationAllowed(type);
         }
     }
-
-    /// <summary>
-    /// Represents an unshared multifile compilation group where types contained in the group are expanded as needed.
-    /// </summary>
-    public class MultiFileLeafCompilationModuleGroup : MultiFileCompilationModuleGroup
-    {
-        public MultiFileLeafCompilationModuleGroup(TypeSystemContext context, IEnumerable<ModuleDesc> compilationModuleSet)
-            : base(context, compilationModuleSet)
-        {
-        }
-
-        public override bool ShouldProduceFullType(TypeDesc type)
-        {
-            // Fully build all shareable types so they will be identical in each module
-            if (EETypeNode.IsTypeNodeShareable(type))
```

Comment:
The problem you were observing would probably also be fixable by additionally conditioning this line on `ConstructedEETypeNode.CreationAllowed(type)`.

If we really wanted to, we could detect the cyclic AssemblyRefs in the driver and use that to drive a decision about which compilation group to create for the leaf (you could just file an issue for that and forget about it). I think the natural treeshaking we get with `MultiFileLeafCompilationModuleGroup` is nice - but I can be convinced otherwise since this is not the "maximize file size savings mode" anyway.